src/node_api.cc

```diff
@@ -3554,3 +3557,299 @@ napi_status napi_run_script(napi_env env,
   *result = v8impl::JsValueFromV8LocalValue(script_result.ToLocalChecked());
   return GET_RETURN_STATUS(env);
 }
+
+struct napi_threadsafe_function__ {
+  uv_async_t async;
+  uv_idle_t idle;
+  uv_mutex_t mutex;
+  uv_cond_t cond;
+  std::queue<void*> queue;
+  size_t refcount;
+  bool is_closing;
+  v8impl::Reference *ref;
+  napi_env env;
+  void *thread_finalize_data;
+  napi_finalize thread_finalize_callback;
+  void* hint;
+  bool idle_running;
+  size_t max_queue_size;
+  napi_threadsafe_function_call_js call_js_cb;
+};
+
+static void napi_threadsafe_function_idle_cb(uv_idle_t* uv_idle) {
+  void* data;
+  bool popped_value = false;
+  bool delete_function = false;
+  size_t size;
+  napi_value js_cb;
+  napi_threadsafe_function ts_fn =
+      node::ContainerOf(&napi_threadsafe_function__::idle, uv_idle);
+
+  uv_mutex_lock(&ts_fn->mutex);
+  size = ts_fn->queue.size();
+  if (size > 0) {
+    data = ts_fn->queue.front();
+    ts_fn->queue.pop();
+    if (size == ts_fn->max_queue_size && ts_fn->max_queue_size > 0) {
+      uv_cond_signal(&ts_fn->cond);
+    }
+    popped_value = true;
+    size--;
+  }
+  if (size == 0 && ts_fn->refcount == 0) {
+    delete_function = true;
+  }
+  uv_mutex_unlock(&ts_fn->mutex);
+
+  v8::HandleScope handle_scope(ts_fn->env->isolate);
+
+  if (popped_value) {
+    js_cb = v8impl::JsValueFromV8LocalValue(ts_fn->ref->Get());
+
+    node::CallbackScope(ts_fn->env->isolate,
+        v8::Local<v8::Object>::Cast(v8impl::V8LocalValueFromJsValue(js_cb)),
+        {0, 0});
+
+    ts_fn->call_js_cb(ts_fn->env, js_cb, ts_fn->hint, data);
+  }
+
+  if (size == 0) {
+    if (uv_idle_stop(&ts_fn->idle) != 0) {
+      napi_throw_error(ts_fn->env, nullptr,
+          "N-API thread-safe function: Failed to stop idle loop");
+      return;
+    }
+    ts_fn->idle_running = false;
+  }
+
+  if (delete_function) {
+    v8impl::Reference::Delete(ts_fn->ref);
+    uv_mutex_destroy(&ts_fn->mutex);
+
+    if (ts_fn->max_queue_size > 0) {
+      uv_cond_destroy(&ts_fn->cond);
+    }
+
+    // It is important that we delete the async first, because the async
+    // callback is the only place from which we awaken the idler. We must avoid
+    // waking the idler because we will have partially destroyed the thread-safe
+    // function by the time the idle handler runs. After all, we have already
+    // destroyed the mutex, the reference, and potentially the condition in the
+    // code above.
+    uv_close(reinterpret_cast<uv_handle_t*>(&ts_fn->async),
+        [] (uv_handle_t* handle) -> void {
+          napi_threadsafe_function ts_fn =
+              node::ContainerOf(&napi_threadsafe_function__::async,
+                reinterpret_cast<uv_async_t*>(handle));
+          uv_close(reinterpret_cast<uv_handle_t*>(&ts_fn->idle),
```

Comment:
These `uv_close` calls should use Node’s `Environment::CloseHandle`, so that we don’t try to shut down the event loop while these callbacks are still pending upon cleanup